src/PABLO/LocalTree.cpp

```diff
@@ -3222,53 +2992,156 @@ namespace bitpit {
      * \param[in] octants list of octants
      * \return Local index of the target octant (=nocts if target Morton not found).
      */
     uint32_t
-    LocalTree::_findMorton(uint64_t Morton, const octvector &octants) const {
-
-        uint32_t nocts = octants.size();
-
-        uint32_t low  = 0;
-        uint32_t high = nocts - 1;
-        while (low <= high) {
-            uint32_t mid = low + ((high - low) / 2);
-            uint64_t mid_morton = octants[mid].computeMorton();
-            if (mid_morton > Morton) {
-                high = mid - 1;
-            } else if (mid_morton < Morton) {
-                low = mid + 1;
-            } else {
-                return mid;
-            }
+    LocalTree::findMorton(uint64_t targetMorton, const octvector &octants) const {
+
+        uint32_t lowerBoundIdx;
+        uint64_t lowerBoundMorton;
+        findMortonLowerBound(targetMorton, octants, &lowerBoundIdx, &lowerBoundMorton);
+
+        uint32_t targetIdx;
+        if (lowerBoundMorton == targetMorton) {
+            targetIdx = lowerBoundIdx;
+        } else {
+            targetIdx = octants.size();
         }
 
-        return nocts;
+        return targetIdx;
     };
 
+    // =================================================================================== //
+    /*! Given a target Morton number and a sorted list of octants, finds the
+     *  index of the first octant whose Morton number does not compare less
+     *  than the target Morton number (in other words, the index of the first
+     *  octant whose Morton number is greater or equal than the target Morton
+     *  number). If the target Morton is greater than the Morton number of the
+     *  last element, the index of the past-the-element element is returned.
+     * \param[in] targetMorton is the Morton index to be found.
+     * \param[in] octants list of octants
+     * \param[out] lowerBoundIdx on output will contain the index of first
+     * octant whose Morton number does not compare less than the target Morton
+     * number. If the target Morton numer is greater than the Morton number of
+     * the last element, the index of the past-the-element element is returned
+     * \param[out] lowerBoundMorton on output will contain the Morton associated
+     * with the lower bound. If the target Morton is greater than the Morton
+     * number of the last element, the maximum finite value representable by
+     * the numeric type is returned
+     */
+    void
+    LocalTree::findMortonLowerBound(uint64_t targetMorton, const octvector &octants, uint32_t *lowerBoundIdx, uint64_t *lowerBoundMorton) const {
+
+        uint32_t nOctants = octants.size();
+
+        uint32_t lowIndex  = 0;
+        uint32_t highIndex = nOctants;
+        uint32_t midIndex  = nOctants;
+        uint64_t midMorton = PABLO::INVALID_MORTON;
+        while (lowIndex < highIndex) {
+            midIndex  = lowIndex + (highIndex - lowIndex) / 2;
```

Comment:
@andrea-iob if I well understand here you're placing always in the worst case. The information on the face (or other entity) through you're searching for the neighbour can be used to fix the portion of the octants container containing the target morton. For mesh with large number of cells I think that the saved time can become considerable.

Reply:
The first step of the binary search is always halfway from the beginning and the end of the container. Previous code was using information on the current octant to adjust the first step of the binary search. However, adjusting the first step of the binary search requires the evaluation of one Morton number and, in terms of computational costs, this is equivalent to one step of binary search. Also, it seems to me, that there were cases where the guess for the initial step provided by the old code was just marginally better than starting halfway in the vector. Therefore, I preferred clean code over a small performance gain. If you think there are cases were the performance gains are noticeable we can re-implement the guess for the initial step later.

Reply:
I've noticed in the old code some performance worsening for large meshes. However, the old version was badly coded, so maybe it was caused by bad implementation. 
Concerning the first Morton evaluation, maybe you don't need it, but the linear index of the current octant could be enough.
Anyway, we have just to remeber, in case of future slowdowns in applications, to check the effect of this initial guess in neighbour research.